blockstack-vm/src/functions/define.rs

```diff
@@ -0,0 +1,132 @@
+                                Err(Error::ExpectedListPairs)
+                            } else {
+                                Ok((&as_vec[0], &as_vec[1]))
+                            }
+                        } else {
+                            Err(Error::ExpectedListPairs)
+                        }
+                    }).collect();
+
+            // step 2: turn into a vec of (name, typesignature) pairs.
+            let key_types: Result<Vec<_>, Error> =
+                (as_pairs?).iter().map(|(name_symbol, type_symbol)| {
+                    let name = match name_symbol {
+                        Atom(ref var) => Ok(var.clone()),
+                        _ => Err(Error::ExpectedListPairs)
+                    }?;
+                    let type_info = match type_symbol {
+                        Atom(ref type_description) => TypeSignature::parse_type_str(type_description),
+                        _ => Err(Error::ExpectedListPairs)
+                    }?;
+                    Ok((name, type_info))
+                }).collect();
+
+            key_types
+        },
+        _ => Err(Error::ExpectedListPairs)
+    }?;
+
+    TupleTypeSignature::new(mapped_key_types)
+}
+
+fn handle_define_map(map_name: &SymbolicExpression,
+                     key_type: &SymbolicExpression,
+                     value_type: &SymbolicExpression) -> Result<DefineResult, Error> {
+    let map_str = match map_name {
+        Atom(ref map_name) => Ok(map_name.clone()),
+        _ => Err(Error::InvalidArguments("Non-name argument to define-map".to_string()))
+    }?;
+
+    let key_type_signature = parse_name_type_pair_list(key_type)?;
+    let value_type_signature = parse_name_type_pair_list(value_type)?;
+
+    Ok(DefineResult::Map(map_str, key_type_signature, value_type_signature))
+}
+
+pub fn evaluate_define(expression: &SymbolicExpression, env: &mut Environment) -> Result<DefineResult, Error> {
+    if let SymbolicExpression::List(elements) = expression {
+        if let Some(Atom(func_name)) = elements.get(0) {
+            return match func_name.as_str() {
+                "define" => {
+                    if elements.len() != 3 {
+                        Err(Error::InvalidArguments("(define ...) requires 2 arguments".to_string()))
+                    } else {
+                        match elements[1] {
+                            Atom(ref variable) => handle_define_variable(variable, &elements[2], env),
+                            AtomValue(ref _value) => Err(Error::InvalidArguments(
+                                "Illegal operation: attempted to re-define a value type.".to_string())),
+                            NamedParameter(ref _value) => Err(Error::InvalidArguments(
+                                "Illegal operation: attempted to re-define a named parameter.".to_string())),
+                            List(ref function_signature) => handle_define_function(&function_signature, &elements[2])
```

Comment:
Is the user allowed to define functions that have reserved names?

Reply:
Right now, it will allow it, though they'll never get called, because `lookup_function` tries the reserved names first,  but it should be an error on `define`